lib/document/stateTransition/validation/structure/validateDocumentsSTStructureFactory.js

```diff
@@ -1,152 +1,171 @@
+    // check data contract exists
+    const dataContract = await dataProvider.fetchDataContract(dataContractId);
 
-    if (!result.isValid()) {
-      return result;
-    }
-
-    if (documentsForDifferentContracts.length > 1) {
+    if (!dataContract) {
       result.addError(
-        new STContainsDocumentsForDifferentDataContractsError(
-          documentsForDifferentContracts,
-        ),
+        new DataContractNotPresentError(dataContractId),
       );
-
-      return result;
     }
 
-    // Fetch Data Contract
-    const [firstRawDocument] = rawStateTransition.documents;
-
-    const dataContractValidationResult = await fetchAndValidateDataContract(firstRawDocument);
-    if (!dataContractValidationResult.isValid()) {
-      result.merge(
-        dataContractValidationResult,
+    const createTransitions = rawStateTransition.transitions
+      .filter(
+        (t) => (t.$action === AbstractDocumentTransition.ACTIONS.CREATE),
       );
 
-      return result;
-    }
-
-    const dataContract = dataContractValidationResult.getData();
-
-    // Validate documents
-    rawStateTransition.documents.forEach((document, index) => {
-      const action = rawStateTransition.actions[index];
-
-      result.merge(
-        validateDocument(document, dataContract, { action }),
+    const replaceTransitions = rawStateTransition.transitions
+      .filter(
+        (t) => (t.$action === AbstractDocumentTransition.ACTIONS.REPLACE),
       );
-    });
-
-    if (!result.isValid()) {
-      return result;
-    }
 
-    // Convert raw documents to Document instances
-    const documents = rawStateTransition.documents.map((rawDocument, index) => {
-      const document = new Document(rawDocument);
+    const deleteTransitions = rawStateTransition.transitions
+      .filter(
+        (t) => (t.$action === AbstractDocumentTransition.ACTIONS.DELETE),
+      );
 
-      document.setAction(rawStateTransition.actions[index]);
+    createTransitions
+      .forEach((rawTransition) => {
+        // validate id generation
+        const documentId = generateDocumentId(
+          dataContractId,
+          ownerId,
+          rawTransition.$type,
+          rawTransition.$entropy,
+        );
+
+        if (rawTransition.$id !== documentId) {
+          result.addError(
+            new InvalidDocumentTransitionIdError(rawTransition),
+          );
+        }
 
-      return document;
-    });
+        // validate entropy
+        if (!entropy.validate(rawTransition.$entropy)) {
+          result.addError(
+            new InvalidDocumentTransitionEntropyError(rawTransition),
+          );
+        }
+      });
+
+    // Validate schema of CREATE and REPLACE transitions
+    createTransitions
+      .concat(replaceTransitions)
+      .forEach((rawTransition) => {
+        // validate document schema
+        const documentSchemaRef = dataContract.getDocumentSchemaRef(rawTransition.$type);
+
+        let enrichedDataContract = enrichDataContractWithBaseSchema(
```

Comment:
We can do that only once

Reply:
@shumkov We can enrich it once, yes. But we also can enrich previous result once again. Why not?